### PR TITLE
Fix registerReverseRoute caching urls

### PR DESCRIPTION
**Pull request by alexdavid**
@charlierudolph `url` needs to be closed over in the reverse route method or it will modify the reference to `url` passed to `registerReverseRoute`.

### CODE REVIEW

**Comment by charlierudolph** [comment]
Hmm, interesting. I find the do a bit odd here. Here are a couple other solutions. Unsure if these are any cleaners or more intuitive.

Solution 1: introduce new variable (I was originally going to do this but couldn't think of a good name at the time)

``` coffee
# Save a function to @reversePaths to get a url back from a route name
registerReverseRoute: ({routeName, url}) ->
  @reversePaths[routeName] = (args...) ->
    resolvedUrl = url
    if typeof args[0] is 'object'
      resolvedUrl = resolvedUrl.replace ":#{k}", v for k, v of args[0]
    else
      resolvedUrl = resolvedUrl.replace /\:[^/]+/, arg for arg in args
    resolvedUrl
```

Solution 2: introduce new function

``` coffee
# Save a function to @reversePaths to get a url back from a route name
registerReverseRoute: ({routeName, url}) ->
  @reversePaths[routeName] = @resolveUrl.bind @, url


# Resolve a url with the given arguments
resolveUrl: (url, args...) ->
  if typeof args[0] is 'object'
    url = url.replace ":#{k}", v for k, v of args[0]
  else
    url = url.replace /\:[^/]+/, arg for arg in args

  url
```

**Comment by alexdavid** [comment]
I agree the `do` is awkward. I like the separate method approach, we could even extract this out into a url building class

**Comment by alexdavid** [comment]
@charlierudolph can you take another look at this when you get a chance?

File: src/exprestive.coffee [hunk 1]
```diff
@@ -89,13 +90,8 @@ class Exprestive
...
   registerReverseRoute: ({routeName, url}) ->
-    @reversePaths[routeName] = (args...) ->
-      if typeof args[0] is 'object'
-        url = url.replace ":#{k}", v for k, v of args[0]
-      else
-        url = url.replace /\:[^/]+/, arg for arg in args
-
-      url
+    formatter = new URLFormatter url
+    @reversePaths[routeName] = (args...) -> formatter.replaceParams args...
```
Comment (charlierudolph) [inline src/exprestive.coffee]: Could this just be 

``` coffee
formatter = new URLFormatter url
@reversePaths[routeName] = formatter.replaceParams
```

With that could then remove the formatter variable

``` coffee
@reversePaths[routeName] = new URLFormatter(url).replaceParams
```

replaceParams will probably need a fat arrow to make that work.

Comment (alexdavid) [inline src/exprestive.coffee]: +1

**Comment by charlierudolph** [comment]
Awesome. :ship:

**Pull request closed without merging**